lazy_tensor_core/lazy_tensor_core/csrc/LazyShapeInference.cpp

```diff
@@ -424,6 +425,137 @@ std::vector<Shape> compute_shape_nll_loss2d_backward(
+  return {Shape(self.scalar_type(), self.sizes().vec())};
+}
+
+std::vector<Shape> compute_shape__adaptive_avg_pool2d(const at::Tensor & self, at::IntArrayRef output_size) {
+  // Checks based on `aten/src/ATen/native/AdaptiveAveragePooling.cpp`
+  // and on `aten/src/ATen/native/cpu/AdaptiveAvgPoolKernel.cpp`
+  TORCH_CHECK(output_size.size() == 2, "adaptive_avg_pool2d: output_size must be 2");
+  TORCH_CHECK(
+      (output_size[0] >= 0 && output_size[1] >= 0),
+      "adaptive_avg_pool2d: elements of output_size must be greater than or equal to 0 ",
+      "but received {", output_size[0], ", ", output_size[1], "}");
+    int64_t ndim = self.ndimension();
+    for (const auto i : c10::irange(1, ndim)) {
+      TORCH_CHECK(self.size(i) > 0,
+        "adaptive_avg_pool2d(): Expected self to have non-zero size for non-batch dimensions, "
+        "but Tensor has sizes ", self.sizes(), " with dimension ", i, " being "
+        "empty");
+    }
+    TORCH_CHECK((ndim == 3 || ndim == 4),
+      "adaptive_avg_pool2d(): Expected 3D or 4D tensor, but got ", self.sizes());
+
+  int64_t channels  = self.size(-3);
+  int64_t output_height = output_size[0];
+  int64_t output_width = output_size[1];
+
+  if (ndim == 3) {
+    return {Shape(self.scalar_type(), {channels, output_height, output_width})};
+  } else {
+    int64_t nbatch = self.size(0);
+    return {Shape(self.scalar_type(), {nbatch, channels, output_height, output_width})};
+  }
+}
+
+std::vector<Shape> compute_shape__adaptive_avg_pool2d_backward(const at::Tensor & grad_output, const at::Tensor & self) {
+    // Checks based on `aten/src/ATen/native/AdaptiveAveragePooling.cpp`
+    int64_t ndim = grad_output.ndimension();
+
+    for (const auto i : c10::irange(1, ndim)) {
+      TORCH_CHECK(grad_output.size(i) > 0,
+        "adaptive_avg_pool2d_backward(): Expected grad_output to have non-zero size for non-batch dimensions, "
+        "but grad_output has sizes ", grad_output.sizes(), " with dimension ", i, " being "
+        "empty");
+    }
+
+    TORCH_CHECK((ndim == 3 || ndim == 4),
+      "adaptive_avg_pool2d_backward(): Expected 3D or 4D tensor, but got ", self.sizes());
+    TORCH_CHECK(self.dtype() == grad_output.dtype(),
+      "expected dtype ", self.dtype(), " for `grad_output` but got dtype ", grad_output.dtype());
+
+    return {Shape(self.scalar_type(), self.sizes().vec())};
+}
+
+std::vector<Shape> compute_shape_glu_backward(const at::Tensor & grad_output, const at::Tensor & self, int64_t dim) {
+  // Checks copied from `aten/src/ATen/native/GatedLinearUnit.cpp`
+  TORCH_CHECK(self.dim() > 0, "glu does not support 0-dimensional tensors");
+  
+  auto wrap_dim = at::maybe_wrap_dim(dim, self.dim());
+  const int64_t nIn = self.size(wrap_dim);
+
+  TORCH_CHECK(nIn % 2 == 0, "Halving dimension must be even, but dimension ",
+              wrap_dim, " is size ", nIn);
+
+  auto sizes = self.sizes().vec();
+  sizes[wrap_dim] = nIn / 2;
```

Comment:
I think this is wrong.  I'd intuitively expect the correct shape for glu_backward to be self.sizes().vec(), since glu's output shape is the same as its input shape (same as other 'activation' or element-wise functions).

Where you probably got confused was in the impl of glu_backward_cpu_out, they actually divide the grad_input tensor into 2 separate halves and do different computations over each half.  I don't actually understand what glu_backward is doing or why it is done in 2 halves.  

What I do see though is that we return `grad_input` which above is resized_as(input), so the output shape should match the input shape regardless of the other stuff.